File: lua/windows/ohmygpu_cmd.lua
local function format_memory(bytes_value)
    local units = {"B", "KB", "MB", "GB", "TB"}
    local i = 1
    while bytes_value >= 1024 and i < #units do
        bytes_value = bytes_value / 1024
        i = i + 1
    end
    return string.format("%.1f %s", bytes_value, units[i])
end

local function run_command(cmd)
    local handle = io.popen(cmd)
    if handle then
        local result = handle:read("*a")
        handle:close()
        return result
    end
    return nil
end

local function get_os_name()
    if jit and jit.os then
        return jit.os
    end

    local sep = package.config:sub(1,1)
    if sep == "\\" then
        return "Windows"
    else
        local f = io.popen("uname -s")
        if f then
            local os = f:read("*l")
            f:close()
            return os
        end
    end
    return "Unknown"
end

local function get_gpu_info_windows()
    local nvidia = run_command('nvidia-smi --query-gpu=name,memory.total,memory.used,utilization.gpu --format=csv,noheader')
    if nvidia and nvidia ~= "" then
        local name, total, used, util = nvidia:match("([^,]+),%s*([^,]+),%s*([^,]+),%s*([^,]+)")
        return {
            name = name,
            total_memory = total,
            used_memory = used,
            utilization = util
        }
    end

    local amd = run_command('rocm-smi --showid --showtemp')
    if amd and amd:find("GPU") then
        return {
            name = "AMD Radeon GPU",
            total_memory = "N/A",
            used_memory = "N/A",
            utilization = "N/A"
        }
    end

    local wmic = run_command('wmic path win32_videocontroller get name,adapterram')
    if wmic then
        for line in wmic:gmatch("[^\r\n]+") do
            local name, mem = line:match("^(.-)%s+(%d+)$")
            if name and mem then
                return {
                    name = name,
                    total_memory = format_memory(tonumber(mem)),
                    used_memory = "N/A",
                    utilization = "N/A"
                }
            end
        end
    end

    return nil
end

local function get_gpu_info()
    local os_name = get_os_name()
    if os_name == "Windows" then
        return get_gpu_info_windows()
    else
        return nil
    end
end

local function print_gpu_info()
    print("\nOH MY GPU:\n")
    local gpu_info = get_gpu_info()
    if gpu_info then
        print(string.format("GPU model:      %s", gpu_info.name))
        print(string.format("Total memory:   %s", gpu_info.total_memory))
        print(string.format("Used memory:    %s", gpu_info.used_memory))
        print(string.format("Utilization:    %s", gpu_info.utilization))
        print("\nGPU is fine.")
    else
        print("[ERROR] GPU not found!")
        print("[ERROR] Make sure GPU drivers are installed.")
    end
end

print_gpu_info()

File: c/windows/ohmygpu_cmd.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void format_memory(long bytes, char *output, size_t size) {
    if (bytes < 1024) {
        snprintf(output, size, "%ld B", bytes);
    } else if (bytes < 1024 * 1024) {
        snprintf(output, size, "%.1f KB", bytes / 1024.0);
    } else if (bytes < 1024 * 1024 * 1024) {
        snprintf(output, size, "%.1f MB", bytes / (1024.0 * 1024));
    } else {
        snprintf(output, size, "%.1f GB", bytes / (1024.0 * 1024 * 1024));
    }
}

void get_gpu_info() {
    FILE *fp;
    char buffer[1024];

    fp = _popen("nvidia-smi --query-gpu=name,memory.total,memory.used,utilization.gpu --format=csv,noheader", "r");
    if (fp) {
        if (fgets(buffer, sizeof(buffer), fp) != NULL) {
            char name[128], total_mem[64], used_mem[64], utilization[64];
            if (sscanf(buffer, "%127[^,], %63[^,], %63[^,], %63[^\n]", name, total_mem, used_mem, utilization) == 4) {
                printf("GPU model:      %s\n", name);
                printf("Total memory:   %s\n", total_mem);
                printf("Used memory:    %s\n", used_mem);
                printf("Utilization:    %s\n", utilization);
                _pclose(fp);
                return;
            }
        }
        _pclose(fp);
    }

    fp = _popen("wmic path win32_VideoController get Name,AdapterRAM", "r");
    if (fp) {
        int first_line = 1;
        while (fgets(buffer, sizeof(buffer), fp)) {
            if (first_line) {
                first_line = 0;
                continue;
            }
            if (strlen(buffer) > 2) {
                char name[128];
                long mem = 0;
                if (sscanf(buffer, "%127[^\r\n0123456789]%ld", name, &mem) >= 1) {
                    char mem_str[64];
                    if (mem > 0) {
                        format_memory(mem, mem_str, sizeof(mem_str));
                    } else {
                        snprintf(mem_str, sizeof(mem_str), "N/A");
                    }
                    printf("GPU model:      %s\n", name);
                    printf("Total memory:   %s\n", mem_str);
                    printf("Used memory:    N/A\n");
                    printf("Utilization:    N/A\n");
                    _pclose(fp);
                    return;
                }
            }
        }
        _pclose(fp);
    }

    printf("[ERROR] GPU not found!\n");
    printf("[ERROR] Make sure GPU drivers are installed.\n");
}

int main() {
    printf("\nOH MY GPU:\n\n");

    get_gpu_info();

    printf("\nPress Enter, for exit...");
    getchar();

    return 0;
}
